packages/eslint-plugin/src/rules/require-object-type-annotations.ts

```diff
@@ -0,0 +1,54 @@
+import type { RuleListener } from '@typescript-eslint/utils/src/ts-eslint';
+import type * as ts from 'typescript';
+
+import * as util from '../util';
+
+export default util.createRule({
+  name: 'require-object-type-annotations',
+  meta: {
+    docs: {
+      description:
+        'Require type annotations for objects where there is no contextual type.',
+      requiresTypeChecking: true,
+      recommended: false,
+    },
+    messages: {
+      forbidden: 'Object is missing type annotation.',
+    },
+    schema: [],
+    type: 'problem',
+  },
+  defaultOptions: [],
```

Comment:
Continuing the thread from here: https://twitter.com/OliverJAsh/status/1570709040765997056

Sometimes we deliberately pass an object into a generic function and we don't want to annotate that object. [`io-ts`](https://github.com/gcanti/io-ts) example:

```ts
const T = t.type({ prop: t.number });
```

I want to add an option whereby users can define a whitelist of function names in which the error will not trigger.

The function name whitelisted by the user could be a member expression or it could just be an identifier:

```ts
ignoreFunctions: [
  // Member expression
  't.type',
  // Identifier
  'foo'
]
```

I initially tried to do this using `whitelist.includes(callee.getText())` but this of course breaks in complex cases such as:

```ts
const T = t
  // comment  
  .type({ prop: t.number });
```

I'm not sure how to check whether the callee matches the whitelist—especially given we don't know whether the whitelisted function is a member expression (e.g. `t.type`) or just an identifier (e.g. `foo`).

Alternatively the rule option could ask for ESQuery selectors and then we check whether the selector matches the callee.

```
excludeCallees: [
  'MemberExpression[object.name="t"][property.name="type"]',
  'Identifier[name="foo"]'
]
```

… but I'm not sure how to check whether a node matches a selector? Are there any examples of this in other rules?

Reply:
#5271 is the current RFC for describing a standardized way to do that. It's tricky; we wouldn't want to allowlist _all_ things named `t` by accident.